lib/nodejs/persistence.js

```diff
@@ -249,21 +249,26 @@ function HandlePersistenceSave( request, response, parsedRequest, segments ) {
         CreateSaveDirectory( parsedRequest[ 'public_path' ], saveName, saveRevision );
         var persistenceData = "";
         var persistenceLength = 0;
-        request.on( 'data', function ( data ) {
-            persistenceData += data;
-            if ( persistenceData.length > 50000000 ) {
-                persistenceData = "";
-                response.writeHead( 413, { 'Content-Type': 'text/plain' } ).end( );
-                request.connection.destroy();
-            }
-        } );
-        request.on( 'end', function( ) {
-            if ( persistenceData.length > 0 ) {
-                var processedPost = querystring.parse( persistenceData );
-                fs.writeFileSync( helpers.JoinPath( './documents', parsedRequest[ 'public_path' ], saveName, 'saveState' + processedPost[ "extension" ] ), processedPost[ "jsonState" ] );
-                fs.writeFileSync( helpers.JoinPath( './documents', parsedRequest[ 'public_path' ], saveName, 'saveState_' + saveRevision + processedPost[ "extension" ] ), processedPost[ "jsonState" ] );
-            }
-        } );
+        if ( typeof request.body === "object" ) {
+            fs.writeFileSync( helpers.JoinPath( './documents', parsedRequest[ 'public_path' ], saveName, 'saveState' + request.body[ "extension" ] ), request.body[ "jsonState" ] );
+            fs.writeFileSync( helpers.JoinPath( './documents', parsedRequest[ 'public_path' ], saveName, 'saveState_' + saveRevision + request.body[ "extension" ] ), request.body[ "jsonState" ] );
+        } else {
+            request.on( 'data', function ( data ) {
+                persistenceData += data;
+                if ( persistenceData.length > 50000000 ) {
```

Comment:
It fails with an appropriate status code: 413 Request Entity Too Large. I don't know how well the client handles it, but this change doesn't add a new error case. The `body-parser` middleware installed at the top does impose a 100 kB limit now that probably needs to be increased.